### DISCUSSION TITLE
using ods in QThread

### DISCUSSION THREAD

**Discussion by Corvette653** [f35f22fan/Ods2#7] [General]
Hi one more time, hope I don't disturb.
Generating big files using your library takes some time, so I decided to use it in new thread, but I have a problem :/
mainwindow.cpp:
```
void MainWindow::on_pushButton_clicked()
{
    WorkerThread *workerThread = new WorkerThread(this);
    connect(workerThread, &WorkerThread::operationDone, workerThread, &WorkerThread::deleteLater);
    workerThread->start();
}

void WorkerThread::run()
{
    auto *book = ods::Book::New();
    ods::AutoDelete<ods::Book*> ad(book);
    emit operationDone();
}
```
mainwindow.h:
```
#ifndef MAINWINDOW_H
#define MAINWINDOW_H

#include <QMainWindow>
#include <QThread>
#include <ods/ods>

QT_BEGIN_NAMESPACE
namespace Ui { class MainWindow; }
QT_END_NAMESPACE

class WorkerThread : public QThread
{
    Q_OBJECT
public:
    explicit WorkerThread(QObject */*parent*/ = nullptr){}
protected:
    virtual void run() override;
signals:
    void operationDone();
};

class MainWindow : public QMainWindow
{
    Q_OBJECT
public:
    MainWindow(QWidget *parent = nullptr);
    ~MainWindow();
private slots:
    void on_pushButton_clicked();
private:
    Ui::MainWindow *ui;
};
#endif // MAINWINDOW_H
```
This code generates an error:
`QThread: Destroyed while thread is still running`

**Comment by f35f22fan** [comment]
I don't know but from looking at it it seems the problem is this line:
 connect(workerThread, &WorkerThread::operationDone, workerThread, &WorkerThread::deleteLater);
You're supposed to delete the thread on "finished" which is a built-in QThread signal.
"operationDone" should be used to tell the rest of your code that the thread finished working and you can access the processed data.

There's a concise example of using QThread in the Qt documentation in the QThread chapter.
https://doc.qt.io/qt-5/qthread.html#details

I personally use pthreads but QThread is the more natural and portable option.

**Reply by Corvette653** [reply]
ehh, dump mistake (like majority of mistakes).
I relied on this example: https://github.com/DeiVadder/QtThreadExample, so I was sure I should have used operationDone.
Sorry, that I disturb you, but this error occured only, when I was using your library, strang though.

**Comment by f35f22fan** [comment]
I forgot to ask, can you send me the code or data so that I can check why it takes long to generate big files? Perhaps you're misusing the API or perhaps I need to fix it.

**Reply by Corvette653** [reply]
could you give me some contact information(e.g. e-mail)?
github doesn't even support uploading ods files

**Reply by f35f22fan** [reply]
f35f22fan AT gmail dot com

**Reply by Corvette653** [reply]
I have sent you a mail fith project in zip
